### PR TITLE
Add ResourceProvider SPI to allow for custom Resources

### PR DESCRIPTION
**Pull request by iNikem**
Partially addresses #1454

### CODE REVIEW

File: sdk/src/main/java/io/opentelemetry/sdk/resources/Resource.java [hunk 1]
```diff
@@ -74,6 +75,14 @@ private static String readVersion() {
     return properties.getProperty("sdk.version");
   }
 
+  private static Resource readResourceFromProviders() {
+    Resource result = Resource.EMPTY;
+    for (ResourceProvider resourceProvider : ServiceLoader.load(ResourceProvider.class)) {
```
Comment (anuraaga) [inline sdk/src/main/java/io/opentelemetry/sdk/resources/Resource.java:80]: We should probably add a property with a CSV list of implementation class name to exclude to allow users to disable resources they don't want to include automatically.

File: sdk/src/main/java/io/opentelemetry/sdk/resources/ResourceProvider.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+ * <p>Resources specified via system properties or environment variables will take precedence
+ * over any value supplied via {@code ResourceProvider}.
+ *
+ * @see EnvAutodetectResource
+ */
+@ThreadSafe
+public abstract class ResourceProvider {
+
+  public Resource create() {
+    final Attributes.Builder attrBuilder = Attributes.newBuilder();
```
Comment (anuraaga) [inline sdk/src/main/java/io/opentelemetry/sdk/resources/ResourceProvider.java:24]: Attributes is immutable guess you don't need to copy?

Comment (iNikem) [reply sdk/src/main/java/io/opentelemetry/sdk/resources/ResourceProvider.java:24]: oi, indeed

File: sdk/src/main/java/io/opentelemetry/sdk/resources/Resource.java (hunk 1 above)
Comment (iNikem) [reply sdk/src/main/java/io/opentelemetry/sdk/resources/Resource.java:80]: Cannot the same be achieved via controlling application classpath?

Comment (anuraaga) [reply sdk/src/main/java/io/opentelemetry/sdk/resources/Resource.java:80]: If every resource provider is in a separate artifact I think so but will that be the case? Not sure - but I sort of figured that SDK will have one extension artifact with a few providers so disabling one of them only by classpath manipulation could be tough.

Comment (iNikem) [reply sdk/src/main/java/io/opentelemetry/sdk/resources/Resource.java:80]: The only use-case for this that I currently know of is for auto-instrumentation to provide `telemetry.auto.version`. Which, btw, I wouldn't want end-users to disable :)

Comment (anuraaga) [reply sdk/src/main/java/io/opentelemetry/sdk/resources/Resource.java:80]: There are many resources we haven't implemented in Java yet, like docker container ID or process. I figure they should also be implemented with this SPI

https://github.com/open-telemetry/opentelemetry-specification/pull/811#discussion_r470931890

We'll probably have a lot of these providers for full data coverage but for startup time or possible privacy reasons presumably they should be disablable and classpath probably won't be enough. In agent we would probably bundle everything and need a property anyways I think?

Comment (iNikem) [reply sdk/src/main/java/io/opentelemetry/sdk/resources/Resource.java:80]: Ok, fair enough.

- Should I do this in this very same PR?
- Should new config property go to `TraceConfig`?

Comment (anuraaga) [reply sdk/src/main/java/io/opentelemetry/sdk/resources/Resource.java:80]: Think follow-up PR is fine.

Probably not `TraceConfig` - we'd have to stop using the static instance, and I guess this doesn't need to be programmatic (can just programmatically set Resource instead) so it's only for control from command line.

Comment (iNikem) [reply sdk/src/main/java/io/opentelemetry/sdk/resources/Resource.java:80]: Added #1551

**Comment by iNikem** [comment]
I will be happy if this gets merged today :) Then I can finish implementing this in instrumentation repo (European) tomorrow.

**Comment by iNikem** [comment]
@jkwatson @tylerbenson @carlosalberto ^^

**Pull request merged**